I/O threads improvements (#4)
martinrvisser/valkey-io.github.io: content/blog/2025-03-27-valkey-8-1-0-rc1.md

### Other Performance Improvements

Several performance improvements were also added to the 8.1 release, here's a list of a few of them:

#### I/O threads improvements

Following up the I/O threads improvements added in 8.0, more operations have been offloaded to the I/O thread pool in the 8.1 release, improving the throughput and latency of some operations.

In the new 8.1 release, TLS connections are now able to efficiently use I/O threads for the several phases of the protocol. The TLS negotiation, which takes non-nigliglibe amount of time to setup new connections, can now be offloaded to the I/O thread pool. This change has improved the rate of accepting new connections by around 300%.

Other sources of overhead in the TLS connection handling were identified, namely the in the calls to `SSL_pending()` and `ERR_clear_error()` functions, which were being called in the main event thread. By offloading these functions to the I/O threads pool, a throughput improvement was achieve in some operations. For instance, it was observed a 10% improvement in `SET` operations throughput, and a 22% improvement in `GET` operations throughput.

> Replication traffic efficiency was also improved in 8.1 by offloading the replication stream reading, and the replica writes to the I/O thread pool. Replica reads were kept in the main thread only to reduce the complexity and because read traffic from replicas is negligible.

Author: ```suggestion
Replication traffic efficiency was also improved in 8.1 by offloading the reading of replication stream on the replicas to the I/O thread pool which means they can serve more read traffic. On the primaries, replication stream writes are now offloaded to the I/O thread pool.
```

Author: @martinrvisser is this change good?
